SCons: Unify tools/target build type configuration

akien-mga opened the pull request:
Implements https://github.com/godotengine/godot-proposals/issues/3371.

TL;DR:
- To build the Godot editor for engine development, use `scons target=editor dev_build=yes`
  * You might also want to add `dev_mode=yes` which is an alias for `verbose=yes warnings=extra werror=yes tests=yes`
- To build the Godot editor to use while developing your game, use `scons target=editor`
  * If you get crashes and want to get a useful stacktrace to report a bug, rebuild with `scons target=editor debug_symbols=yes`
- Both the intermediate object files (`.o`, `.a`, `.obj`, `.lib`) and the Godot binary in bin are changing names, so you might want to clean your Git checkout to remove the old ones to save disk space, and adjust any scripts/config you might have that runs the Godot binary to use the new name.

### New `target` presets

The `tools` option is removed and `target` changes to use three new presets,
which match the builds users are familiar with. These targets control the
default optimization level and enable editor-specific and debugging code:

- `editor`: Replaces `tools=yes target=release_debug`.
  * Defines: `TOOLS_ENABLED`, `DEBUG_ENABLED`, `-O2`/`/O2`
- `template_debug`: Replaces `tools=no target=release_debug`.
  * Defines: `DEBUG_ENABLED`, `-O2`/`/O2`
- `template_release`: Replaces `tools=no target=release`.
  * Defines: `-O3`/`/O2`

### New `dev_build` option

The previous `target=debug` is now replaced by a separate `dev_build=yes`
option, which can be used in combination with either of the three targets,
and changes the following:

- `dev_build`: Defines `DEV_ENABLED`, disables optimization (`-O0`/`/0d`),
  enables generating debug symbols, does not define `NDEBUG` so `assert()`
  works in thirdparty libraries, adds a `.dev` suffix to the binary name.

Note: Unlike previously, `dev_build` defaults to off so that users who
compile Godot from source get an optimized and small build by default.
Engine contributors should now set `dev_build=yes` in their build scripts or
IDE configuration manually.

### Changed binary names

The name of generated binaries and object files are changed too, to follow
this format:

`godot.<platform>.<target>[.dev][.double].<arch>[.<extra_suffix>][.<ext>]`

For example:
- `godot.linuxbsd.editor.dev.arm64`
- `godot.windows.template_release.double.x86_64.mono.exe`

Be sure to update your links/scripts/IDE config accordingly.

### More flexible `optimize` and `debug_symbols` options

The optimization level and whether to generate debug symbols can be further
specified with the `optimize` and `debug_symbols` options. So the default
values listed above for the various `target` and `dev_build` combinations
are indicative and can be replaced when compiling, e.g.:

`scons p=linuxbsd target=template_debug dev_build=yes optimize=debug`
will make a "debug" export template with dev-only code enabled, `-Og`
optimization level for GCC/Clang, and debug symbols. Perfect for debugging
complex crashes at runtime in an exported project.

-----

The above description was updated to match the latest state of this PR.

<details>
<summary>Original description and comments</summary>

There's a lot to discuss there, I'll try to make a list of the main point.

As described in the proposal, this PR unifies the current `tools` and `target` combinations in a single `target` enum, with the intention of making the valid build targets clearer, and to clarify the confusion around the meaning of the word "debug".

The targets implemented here are therefore:
```
target:
    editor:
        editor: yes (TOOLS_ENABLED)
        debug-features: yes (DEBUG_ENABLED)
        dev-features: no
        optimize: speed-debug (-O2)
        debug_symbols: yes (-g2)
    editor-dev:
        editor: yes (TOOLS_ENABLED)
        debug-features: yes (DEBUG_ENABLED)
        dev-features: yes (DEV_ENABLED)
        optimize: debug (-Og)
        debug_symbols: yes (-g3)
    template-release:
        editor: no
        debug-features: no
        dev-features: no
        optimize: speed (-O3) / size (-Os) on Web
        debug_symbols: no
    template-debug:
        editor: no
        debug-features: yes (DEBUG_ENABLED)
        optimize: speed-debug (-O2) / size (-Os) on Web
        debug_symbols: yes (-g2)
    template-dev:
        editor: no
        debug-features: yes (DEBUG_ENABLED)
        dev-features: yes (DEV_ENABLED)
        optimize: debug (-Og)
        debug_symbols: yes (-g3)
```

### Notes

- The new `target` mostly governs three things:
  * Editor or export template
  * Debugging features (always true for editor, true for "debug" export template)
  * Dev build with full debugging symbols (`-g3 -Og`)
    - This one could possibly be split into a separate `dev_build=yes` option, which would bring down the `target` enum to `"editor", "template-release", "template-debug"`, which are effectively what we ship. Thoughts? This idea is growing on me as I type it.
- `optimize` and `debug_symbols` can be overridden for any target.
   * So if you want, you can build `target=editor-dev optimize=size debug_symbols=no` or something. It doesn't make much sense per se but the flexibility can be useful in other situations like `target=template-release optimize=speed-debug debug_symbols=yes`. This is not exactly the same as `template-debug` as the debug features are still off - so it can be pertinent to debug an issue that only happens in release export templates and may be linked to not have `DEBUG_ENABLED` code.
- The object suffix was changed from the previous mix of `opt`, `tools` and `debug` to just use `target`, so the build objects and final binary names are changing. E.g. `scons p=linuxbsd target=editor-dev arch=x86_64` will produce `bin/godot.linuxbsd.editor-dev.x86_64`.
- The default option for `scons` without target is now `target=editor`, which will imply `optimize=speed-debug debug_symbols=yes`
  * This is different to the current default which is `tools=yes target=debug debug_symbols=yes` which means a debug-optimized (`-Og`) (or actually one some platforms it would be unoptimized `-O0`) build with `DEV_ENABLED`.
  * Basically this makes the default more end-user friendly (they get a usable editor build for using Godot), and contributors will have to specify `target=editor-dev` to get back the previous behavior.
  * The default values for `optimize` and `debug_symbols` in the SCons help are not really accurate as they eventually depend on the `target`. I could add `auto` as a default to clarify this, though this can be a bit tricky to handle properly.
- Compiler and linker flags for the different `optimize` and `debug_symbols` levels are now unified in `SConstruct`, right after `configure` (needs to be after for `env.msvc` to be valid).
  * This should be refactored eventually to reduce the size of the SConstruct and make the code more readable. Could be in a dedicated function / SCons `Tool`.
  * All platforms used to do their own handling of flags for all targets, I did my best to unify what can be but there are some platform-specific tweaks that remain in the various `detect.py`, which should be assessed. There are also some default values changed which I'll point out in review comments.
- `godot-cpp` would need to be ported to support something similar probably. Whether or not it can/should might also inform us on whether the set of target presets proposed here is actually the best design or not.
  * I didn't port the GDExtension compatibility build script for TextServers, it does some reimplementation of the `SHLIBSUFFIX`
- We have a mess of platform-specific defines of `DEBUG=1`, `DEBUG:FULL` `_DEBUG`, `NDEBUG`... this needs a thorough review and unification to make sure we do the right thing in the right situation. Notably, it used to be tied to either using `debug_symbols=yes` or `target=debug` but I'm not sure it's a good idea for all of these.
  * See #66303.
- All these configuration options could possibly be handled a bit more elegantly with e.g. a Python class that keeps track of this data and can be accessed from `SCsub`s via the env, e.g. `env.target.editor_build` / `env.target.template_build` / `env.target.debug_features`, etc.
- Some platforms (Web and iOS) used to set their optimization flags both for the compiler and the linker (`CCFLAGS` and `LINKFLAGS`). I don't know why, but I guess the `LINKFLAGS` lines can be safely dropped?
- Clang based platforms (Android, iOS and macOS) all defined `-ftree-vectorize` for both speed and size optimizations, and `-fomit-frame-pointer` for speed optimizations. This is apparently supported by GCC too, and possibly by Emscripten (LLVM based too), so should we define them consistently for all platforms? On MSVC they're somewhat part of `/O1`/`/O2` already by way of `/Oy`.
  * See #66296 and #66297.

</details>

akien-mga committed:
SCons: Unify tools/target build type configuration

Implements https://github.com/godotengine/godot-proposals/issues/3371.

New `target` presets
====================

The `tools` option is removed and `target` changes to use three new presets,
which match the builds users are familiar with. These targets control the
default optimization level and enable editor-specific and debugging code:

- `editor`: Replaces `tools=yes target=release_debug`.
  * Defines: `TOOLS_ENABLED`, `DEBUG_ENABLED`, `-O2`/`/O2`
- `template_debug`: Replaces `tools=no target=release_debug`.
  * Defines: `DEBUG_ENABLED`, `-O2`/`/O2`
- `template_release`: Replaces `tools=no target=release`.
  * Defines: `-O3`/`/O2`

New `dev_build` option
======================

The previous `target=debug` is now replaced by a separate `dev_build=yes`
option, which can be used in combination with either of the three targets,
and changes the following:

- `dev_build`: Defines `DEV_ENABLED`, disables optimization (`-O0`/`/0d`),
  enables generating debug symbols, does not define `NDEBUG` so `assert()`
  works in thirdparty libraries, adds a `.dev` suffix to the binary name.

Note: Unlike previously, `dev_build` defaults to off so that users who
compile Godot from source get an optimized and small build by default.
Engine contributors should now set `dev_build=yes` in their build scripts or
IDE configuration manually.

Changed binary names
====================

The name of generated binaries and object files are changed too, to follow
this format:

`godot.<platform>.<target>[.dev][.double].<arch>[.<extra_suffix>][.<ext>]`

For example:
- `godot.linuxbsd.editor.dev.arm64`
- `godot.windows.template_release.double.x86_64.mono.exe`

Be sure to update your links/scripts/IDE config accordingly.

More flexible `optimize` and `debug_symbols` options
====================================================

The optimization level and whether to generate debug symbols can be further
specified with the `optimize` and `debug_symbols` options. So the default
values listed above for the various `target` and `dev_build` combinations
are indicative and can be replaced when compiling, e.g.:

`scons p=linuxbsd target=template_debug dev_build=yes optimize=debug`
will make a "debug" export template with dev-only code enabled, `-Og`
optimization level for GCC/Clang, and debug symbols. Perfect for debugging
complex crashes at runtime in an exported project.

bruvzg commented on platform/macos/detect.py:
```diff
@@ -86,27 +84,15 @@ def configure(env):
+    # TODO: Re-evaluate the need for this / streamline with common config.
+    if env["target"] == "template-release":
+        if env["optimize"].startswith("speed"):
+            env.Prepend(CCFLAGS=["-fomit-frame-pointer", "-ftree-vectorize"])
```
These options are not macOS or clang specific, so probably should be set everywhere (`/Oy` for MSVC).

With exception to `-no_deduplicate` (LD64 flag).

akien-mga replied:
Indeed, currently we seem to set `-fomit-frame-pointer -ftree-vectorize` only for Clang builds: Android, iOS, macOS.
I don't know why / if there's a problem using those with GCC. I guess we could try them out in a separate PR so that we don't have to deal with changing this optimization in this massive refactor.

`/Oy` for MSVC seems implied by both `/O1` and `/O2` so I guess it's not needed: https://learn.microsoft.com/en-us/cpp/build/reference/oy-frame-pointer-omission?view=msvc-170

akien-mga replied:
I researched this a bit in #66296, and found that we don't need these flags at all, so #66297 removes them.

bruvzg commented on platform/linuxbsd/detect.py:
```diff
@@ -69,26 +67,8 @@ def configure(env):
-    elif env["target"] == "debug":
-        env.Prepend(CCFLAGS=["-g3"])
+    # TODO: Re-evaluate the need for this / streamline with common config.
+    if env["target"].endswith("-dev"):
```
Not sure why it's here. Isn't it to make executable symbols visible to the libraries?

akien-mga replied:
Yeah we noticed this discrepancy a while ago in March on Rocket.Chat, discussing it with @raulsntos and @neikeq: https://chat.godotengine.org/channel/devel?msg=2Dkxu3KsP2TZEwRNn

It seems it was added by @marcelofg55 in #11061, presumably for the Linux crash handler. I don't know if it's actually needed though. I can try crashing Godot without it and see what happens.

It also used to be used by the Mono module (on all targets, which would actually prevent removing unused symbols, explaining a part of the size difference between Mono and non-Mono builds in 3.x) but it's not needed anymore with .NET.

akien-mga replied:
So I gave it a try, and indeed without `-rdynamic` the crash handler doesn't retrieve the symbols properly.

With `-rdynamic`:
```
================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.0.beta.custom_build (8e14f9ba21725a9445f3979742f2fc87c8a7b463)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[1] /lib64/libc.so.6(+0x36970) [0x7f4f2fa7d970] (??:0)
[2] core_bind::OS::crash(String const&) (/home/akien/Projects/godot/godot.git/core/core_bind.cpp:232)
[3] void call_with_ptr_args_helper<__UnexistingClass, String const&, 0ul>(__UnexistingClass*, void (__UnexistingClass::*)(String const&), void const**, IndexSequence<0ul>) (/home/akien/Projects/godot/godot.git/./core/variant/binder_common.h:283 (discriminator 8))
[4] void call_with_ptr_args<__UnexistingClass, String const&>(__UnexistingClass*, void (__UnexistingClass::*)(String const&), void const**) (/home/akien/Projects/godot/godot.git/./core/variant/binder_common.h:532)
[5] MethodBindT<String const&>::ptrcall(Object*, void const**, void*) (/home/akien/Projects/godot/godot.git/./core/object/method_bind.h:331)
[6] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (/home/akien/Projects/godot/godot.git/modules/gdscript/gdscript_vm.cpp:1965)
[7] GDScript::_create_instance(Variant const**, int, Object*, bool, Callable::CallError&) (/home/akien/Projects/godot/godot.git/modules/gdscript/gdscript.cpp:175)
[8] GDScript::instance_create(Object*) (/home/akien/Projects/godot/godot.git/modules/gdscript/gdscript.cpp:406)
[9] Object::set_script(Variant const&) (/home/akien/Projects/godot/godot.git/core/object/object.cpp:843)
[10] MainLoop::initialize() (/home/akien/Projects/godot/godot.git/core/os/main_loop.cpp:61 (discriminator 4))
[11] OS_LinuxBSD::run() (/home/akien/Projects/godot/godot.git/platform/linuxbsd/os_linuxbsd.cpp:571)
[12] ./bin/godot.linuxbsd.tools.x86_64.rdynamic(main+0x146) [0x2460eec] (/home/akien/Projects/godot/godot.git/platform/linuxbsd/godot_linuxbsd.cpp:74)
[13] /lib64/libc.so.6(+0x23677) [0x7f4f2fa6a677] (??:0)
[14] /lib64/libc.so.6(__libc_start_main+0x85) [0x7f4f2fa6a735] (??:0)
[15] ./bin/godot.linuxbsd.tools.x86_64.rdynamic(_start+0x21) [0x2460ce1] (??:?)
-- END OF BACKTRACE --
================================================================
```

Without `-rdynamic`:
```
================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.0.beta.custom_build (8e14f9ba21725a9445f3979742f2fc87c8a7b463)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[1] /lib64/libc.so.6(+0x36970) [0x7fb97e2c1970] (??:0)
[2] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0x43cc3d7] (/home/akien/Projects/godot/godot.git/core/core_bind.cpp:232)
[3] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0x756b14] (/home/akien/Projects/godot/godot.git/./core/variant/binder_common.h:283 (discriminator 8))
[4] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0x756737] (/home/akien/Projects/godot/godot.git/./core/variant/binder_common.h:532)
[5] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0x7563d0] (/home/akien/Projects/godot/godot.git/./core/object/method_bind.h:331)
[6] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0xc647a4] (/home/akien/Projects/godot/godot.git/modules/gdscript/gdscript_vm.cpp:1965)
[7] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0xb0ad5f] (/home/akien/Projects/godot/godot.git/modules/gdscript/gdscript.cpp:175)
[8] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0xb0c41e] (/home/akien/Projects/godot/godot.git/modules/gdscript/gdscript.cpp:406)
[9] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0x49b029e] (/home/akien/Projects/godot/godot.git/core/object/object.cpp:843)
[10] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0x4479a86] (/home/akien/Projects/godot/godot.git/core/os/main_loop.cpp:61 (discriminator 4))
[11] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0x20fd63] (/home/akien/Projects/godot/godot.git/platform/linuxbsd/os_linuxbsd.cpp:571)
[12] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0x209eec] (/home/akien/Projects/godot/godot.git/platform/linuxbsd/godot_linuxbsd.cpp:74)
[13] /lib64/libc.so.6(+0x23677) [0x7fb97e2ae677] (??:0)
[14] /lib64/libc.so.6(__libc_start_main+0x85) [0x7fb97e2ae735] (??:0)
[15] ./bin/godot.linuxbsd.tools.x86_64.no_rdynamic() [0x209ce1] (??:?)
-- END OF BACKTRACE --
================================================================
```

Size difference:
```
-rwxr-xr-x 1 akien akien 697M Sep 22 13:41 godot.linuxbsd.tools.x86_64.no_rdynamic*
-rwxr-xr-x 1 akien akien 731M Sep 22 13:41 godot.linuxbsd.tools.x86_64.rdynamic*
```

However `gdb` has no problem getting the stacktrace from the build without `-rdynamic` so this might be a problem with our crash handler. It does imply that even with debug symbols, our `release_debug` builds currently wouldn't give a meaningful backtrace.

```
(gdb) bt
#0  0x00000000043cc3d7 in core_bind::OS::crash (this=0x8eb3ee0, p_message=...) at core/core_bind.cpp:232
#1  0x0000000000756b14 in call_with_ptr_args_helper<__UnexistingClass, String const&, 0ul> (p_instance=0x8eb3ee0, 
    p_method=(void (__UnexistingClass::*)(__UnexistingClass * const, const String &)) 0x43cc392 <core_bind::OS::crash(String const&)>, p_args=0x7fffffff9a20) at ./core/variant/binder_common.h:283
#2  0x0000000000756737 in call_with_ptr_args<__UnexistingClass, String const&> (p_instance=0x8eb3ee0, 
    p_method=(void (__UnexistingClass::*)(__UnexistingClass * const, const String &)) 0x43cc392 <core_bind::OS::crash(String const&)>, p_args=0x7fffffff9a20) at ./core/variant/binder_common.h:531
#3  0x00000000007563d0 in MethodBindT<String const&>::ptrcall (this=0x8eb4c30, p_object=0x8eb3ee0, p_args=0x7fffffff9a20, r_ret=0x0) at ./core/object/method_bind.h:329
#4  0x0000000000c647a4 in GDScriptFunction::call (this=0xaabc110, p_instance=0xb7736a0, p_args=0x0, p_argcount=0, r_err=..., p_state=0x0) at modules/gdscript/gdscript_vm.cpp:1962
#5  0x0000000000b0ad5f in GDScript::_create_instance (this=0xb773e20, p_args=0x0, p_argcount=0, p_owner=0x7fffac002d70, p_is_ref_counted=false, r_error=...) at modules/gdscript/gdscript.cpp:175
#6  0x0000000000b0c41e in GDScript::instance_create (this=0xb773e20, p_this=0x7fffac002d70) at modules/gdscript/gdscript.cpp:406
#7  0x00000000049b029e in Object::set_script (this=0x7fffac002d70, p_script=...) at core/object/object.cpp:843
#8  0x0000000004479a86 in MainLoop::initialize (this=0x7fffac002d70) at core/os/main_loop.cpp:61
#9  0x000000000020fd63 in OS_LinuxBSD::run (this=0x7fffffffcff0) at platform/linuxbsd/os_linuxbsd.cpp:563
#10 0x0000000000209eec in main (argc=3, argv=0x7fffffffd548) at platform/linuxbsd/godot_linuxbsd.cpp:72
```

akien-mga replied:
For now I'm adding a comment to clarify that, but we might be able to get rid of the flag by reworking the crash handler.

akien-mga commented on SConstruct:
```diff
@@ -475,24 +515,55 @@ if selected_platform in platform_list:
         env["verbose"] = methods.get_cmdline_bool("verbose", True)
         env["warnings"] = ARGUMENTS.get("warnings", "extra")
         env["werror"] = methods.get_cmdline_bool("werror", True)
-        if env["tools"]:
-            env["tests"] = methods.get_cmdline_bool("tests", True)
+        env["tests"] = methods.get_cmdline_bool("tests", True)
```
This changes behavior, tests will now be compiled also for template builds with the `dev=yes` convenience shortcut.
We already have tests working on non-editor builds so there's no reason to exclude them when full-blown dev mode is requested.

akien-mga commented on SConstruct:
```diff
@@ -475,24 +515,55 @@ if selected_platform in platform_list:
     # are actually handled to change compile options, etc.
     detect.configure(env)
 
+    # Set optimize and debug_symbols flags.
+    # Needs to happen after configure to have `env.msvc` defined.
```
@Faless I started from https://github.com/godotengine/godot-cpp/blob/master/tools/targets.py but as I started syncing with the platform-specific `detect.py` I had to change some things which should likely be unified further in godot-cpp:

- Removed `0` to `3` flags as `3` isn't valid for MSVC, and IMO the use for these is limited. I think users can specify what they need with `CCFLAGS` if they need to override what we have (we could add `optimize=custom` for this if need be, to make sure we don't end up adding `-O0` from `optimize=none` after user-specified `CCFLAGS`).
- Removed `auto` as the logic is more complex than this, instead we handle it in `SConstruct` when handling the targets.
- Removed forcing debug symbols on `-dev` (ex `debug`) builds. There's a sane default now, but I think we should still give precedence to the `debug_symbols` option if specified.
- MSVC:
  * Removed `/Z7` for builds without debug symbols. Copy paste mistake I believe?
  * Added `/FS` for `debug_symbols=yes` as this was used in `platform/windows/detect.py`. Didn't assess what it is/does.
  * `size` optimization: the proper flag for this is `/O1`: https://learn.microsoft.com/en-us/cpp/build/reference/o1-o2-minimize-size-maximize-speed?view=msvc-170. `/Os` is a subset of `/O1` but the equivalent to GCC `-Os` is `/O1` as far as I know (this also makes a case for not exposing `0` - `3` flags as `-O1` and `/O1` are quite different).
  * Added `env.Append(LINKFLAGS=["/OPT:REF"])` which was in the Windows `detect.py`.
- GCC/Clang:
  * Added `speed-debug` for `-O2` on what used to be `release_debug` builds, as that's what we used there (apparently to keep potential backtraces somewhat usable).

Faless replied:
I think all the changes are good, and I agree having a `optmize=custom` that requires you to set CCFLAGS seems reasonable.

`/FS` seems to be a partial workaround to windows having issues with multiprocessing.
They say it helps though it doesn't actually fix all the problems: https://learn.microsoft.com/en-us/cpp/build/reference/fs-force-synchronous-pdb-writes?view=msvc-170
Seems safe to use.

akien-mga replied:
I added `optimize=custom` in the latest iteration.

akien-mga commented on platform/android/detect.py:
```diff
@@ -116,19 +116,14 @@ def configure(env):
-            # when using `target=release_debug`.
-            opt = "-O3" if env["target"] == "release" else "-O2"
-            env.Append(CCFLAGS=[opt, "-fomit-frame-pointer"])
-        elif env["optimize"] == "size":  # optimize for size
-            env.Append(CCFLAGS=["-Oz"])
```
A couple differences for Android:

- `optimize=size` used to be `-Oz`, now it's `-Os` like for all other platforms. `-Oz` is apparently a more aggressive size optimization flag, but for the Web build we have this info:
```
# -Os reduces file size by around 5 MiB over -O3. -Oz only saves about
# 100 KiB over -Os, which does not justify the negative impact on
# run-time performance.
```

- `target=debug` used to force no optimization (`-O0`), now it uses debug optimization (`-Og`) like other platforms, which is a subset of `-O1` which does not hinder debugging.

akien-mga commented on platform/android/detect.py:
```diff
@@ -116,19 +116,14 @@ def configure(env):
-        env.Append(LINKFLAGS=["-O0"])
-        env.Append(CCFLAGS=["-O0", "-g", "-fno-limit-debug-info"])
+    else:
+        env.Append(CCFLAGS=["-fno-limit-debug-info"])
         env.Append(CPPDEFINES=["_DEBUG"])
         env.Append(CPPFLAGS=["-UNDEBUG"])
```
Some of these defines seem to be overkill for Android:
- `-fno-limit-debug-info` isn't defined on any other platform so not sure why it's needed here. Seems like it was added a long time ago by @RandomShaper in #6958, do you remember why?
- Undefining `NDEBUG` shouldn't be needed since we don't define it. But most importantly we want to avoid exceptions and AFAIK this controls exception behavior in `assert` so I'm not sure why we specifically want to re-enable it. Was also added in #6958.
- `_DEBUG` AFAIK is a MSVC specific define used for CRT debugging, so it makes no sense for Android. It's been there since Godot was open sourced, can likely be safely removed as it was likely an oversight.

RandomShaper replied:
In Android I tried to match the flags set by the NDK as close as possible. However, I understand that some may not be technically needed and the NDK is adding them to support Android tooling (i.e., Android Studio). It's hard to tell. Maybe @m4gr3d has some insight on how convenient it'd be to remove it.

akien-mga replied:
Cleaning this up in #66303.

akien-mga commented on platform/ios/detect.py:
```diff
@@ -54,20 +54,15 @@ def configure(env):
+        elif env["optimize"] == "size":
+            env.Append(CCFLAGS=["-ftree-vectorize"])
+    else:
+        env.Append(CCFLAGS=["-gdwarf-2"])
         env.Append(CPPDEFINES=["_DEBUG", ("DEBUG", 1)])
```
@bruvzg
- Does it make a difference for iOS if we use `-g` or `-gdwarf-2`? If not we can remove that first line.
- Worth noting that the change will also make iOS `template-dev` build (`optimize=debug`) use `-Og` like other platforms instead of `-O0` like before.
- `_DEBUG` is likely unnecessary as this is a MSVC specific macro.
- Not sure what `DEBUG=1` does and why this is only defined on iOS.

bruvzg replied:
Macros are probably just copy-pasted for other platform, I do not think it will do anything.

`gdwarf-2` is likely also unnecessary, probably for using debuggers on old macOS versions (IIRC, default for current clang is drawf-4, and Xcode on macOS older than 10.10 do not support it).

akien-mga replied:
Changed it in #66303.

akien-mga commented on platform/ios/detect.py:
```diff
@@ -146,12 +141,10 @@ def configure(env):
         env.Append(ASFLAGS=["-arch", "arm64"])
         env.Append(CPPDEFINES=["NEED_LONG_INT"])
 
-    # Disable exceptions on non-tools (template) builds
-    if not env["tools"]:
```
I removed the check since we don't support editor build on iOS for now. Could be readded once it's needed.

akien-mga commented on platform/web/detect.py:
```diff
@@ -77,26 +77,18 @@ def configure(env):
-        if env["optimize"] == "size":
-            env.Append(CCFLAGS=["-Os"])
-            env.Append(LINKFLAGS=["-Os"])
-        elif env["optimize"] == "speed":
-            env.Append(CCFLAGS=["-O3"])
-            env.Append(LINKFLAGS=["-O3"])
-
-        if env["target"] == "release_debug":
-            # Retain function names for backtraces at the cost of file size.
-            env.Append(LINKFLAGS=["--profiling-funcs"])
-    else:  # "debug"
-        env.Append(CCFLAGS=["-O1", "-g"])
-        env.Append(LINKFLAGS=["-O1", "-g"])
```
@Faless By unifying the code with other platforms a couple changes were made here:
- `optimize=speed` now uses `-O3` for `template-release` but `-O2` for `template-debug`/`editor`, like on other platforms. We optimize for size by default anyway so I figured this isn't a big deal?
- `-dev` builds (ex `target=debug`) default to `debug` optimization level i.e. `-Og`, instead of `-O1` as was set here. It seemed to me the change is fine as `debug` is not something we ship, and `-Og` is supposed to be fairly close to `-O1`, with the exception of those optimizations that hinder debugging.

akien-mga commented on platform/web/detect.py:
```diff
@@ -77,26 +77,18 @@ def configure(env):
-        env.Append(LINKFLAGS=["-O1", "-g"])
+
+    # TODO: Re-evaluate the need for this / streamline with common config.
+    if not env["target"].endswith("-dev"):
+        # Retain function names for backtraces at the cost of file size.
+        env.Append(LINKFLAGS=["--profiling-funcs"])
```
Noticed an unintended logic change here, this used to be added only for `release_debug` so it should check `template-debug` or `editor`.

akien-mga replied:
Fixed.

akien-mga commented on platform/macos/detect.py:
```diff
@@ -86,27 +84,15 @@ def configure(env):
+            env.Prepend(CCFLAGS=["-fomit-frame-pointer", "-ftree-vectorize"])
+        elif env["optimize"] == "size":
+            env.Prepend(CCFLAGS=["-ftree-vectorize"])
         if env["arch"] != "arm64":
             env.Prepend(CCFLAGS=["-msse2"])
```
Our handling of `SSE2` is a bit arbitrary. We seem to forcefully define it for Windows MinGW and macOS x86_64 with `target=release` specifically, and not on any other platform:
```
$ rg msse2
platform/windows/detect.py
496:        env.Append(CCFLAGS=["-msse2"])

platform/macos/detect.py
95:            env.Prepend(CCFLAGS=["-msse2"])

modules/raycast/SCsub
70:            env_raycast.Append(CPPFLAGS=["-msse2", "-mxsave"])
```

akien-mga commented on platform/web/emscripten_helpers.py:
```diff
@@ -38,7 +38,7 @@ def create_engine_file(env, target, source, externs):
 
 def create_template_zip(env, js, wasm, worker, side):
-    binary_name = "godot.tools" if env["tools"] else "godot"
+    binary_name = "godot.editor" if env.editor_build else "godot"
```
@Faless Worth noting that I changed this too while at it. I've changed all cases of `"\.tools"` but I'm not sure that covers everything for the Web platform.

akien-mga commented:
I noticed this size difference between a `tools=yes target=debug` build of the current `master` and `target=editor-dev` from this PR (which should be equivalent). I'll check what explains it.
```
-rwxr-xr-x 1 akien akien 935M Sep 22 09:25 godot.linuxbsd.editor-dev.x86_64*
-rwxr-xr-x 1 akien akien 697M Sep 22 13:41 godot.linuxbsd.tools.x86_64*
```

neikeq commented:
> Edit: The difference is that for LinuxBSD we didn't set `-Og` for `debug` builds. Apparently `-Og` makes bigger binaries than no optimization / `-O0`? That's weird.

The size difference doesn't sound weird to me. What confuses me is why we weren't using `-Og` before.

neikeq commented on SConstruct:
```diff
@@ -475,24 +515,55 @@ if selected_platform in platform_list:
+
+        if env["optimize"] == "speed":
+            env.Append(CCFLAGS=["-O3"])
+        # `-O2` is friendlier to debuggers than `-O3`, leading to better crash backtraces.
+        elif env["optimize"] == "speed-debug":
```
A better name could be `speed-trace`.

neikeq commented:
> The size difference doesn't sound weird to me.

I'm referring to this in particular:

```
-Og 

... It is a better choice than -O0 for producing debuggable code because some compiler passes that collect debug information are disabled at -O0. 
```

I wouldn't trust that it's better for debugging, though. There have been problems in the past, like `-Og` optimizing out variables: https://stackoverflow.com/questions/31435771/variables-optimized-out-with-g-and-the-og-option#comment80796148_37830258. Perhaps that's why we are using `-O0` instead.

aaronfranke commented on SConstruct:
```diff
@@ -376,23 +384,55 @@ env_base.Prepend(CPPPATH=["#"])
+
+target = env_base["target"]
+env_base.editor_build = target.startswith("editor")
+env_base.dev_build = False
```
For consistency, maybe this?

```suggestion
env_base.dev_build = target.endswith("dev")
```

akien-mga replied:
That's what I did initially but then I changed my mind, I don't like relying on arbitrary substrings to much unless we have a very clear pattern that both Godot and derivative projects like godot-cpp would follow.

I think that's another argument for making the `dev_build` aspect a separate SCons option.

aaronfranke commented on SConstruct:
```diff
@@ -376,23 +384,55 @@ env_base.Prepend(CPPPATH=["#"])
+target = env_base["target"]
+env_base.editor_build = target.startswith("editor")
+env_base.dev_build = False
+debug_features = False
+opt_level = ""
+debug_symbols = False
```
A bit more of a stretch, but we could also do that here:

```suggestion
debug_features = not target.endswith("release")
opt_level = ""
debug_symbols = not target.endswith("release")
```

By the way, it looks like currently these flags are always set together, even though they mean different things.

akien-mga replied:
Yeah or `target != "template-release"`. But same as above, I eventually favored being explicit in the `if`/`elif` that handles each target instead of trying to rely on substrings.

> By the way, it looks like currently these flags are always set together, even though they mean different things.

Indeed, but that's just by chance. `debug_symbols` could really take any value, and is configurable from the command line (and can be stripped anyway after the fact), while `debug_features` is fixed and should not be made configurable (it's basically the only difference between `template-release` and `template-debug`, aside from a different default - but configurable - optimization level).

akien-mga replied:
In the end with the new `dev_build` option I refactored this to be much better IMO.

aaronfranke reviewed (approved):
Looks really good overall.

Faless reviewed (approved):
Really good work.
The only remark I would like to add is about:

> This one could possibly be split into a separate dev_build=yes option, which would bring down the target enum to "editor", "template-release", "template-debug", which are effectively what we ship. Thoughts? This idea is growing on me as I type it.

**I wholeheartedly support doing this** to avoid further confusion among the users.
And I think `dev_build=yes` should simply add the `DEV_ENABLED` flag and set default values for debug symbols and optimization levels so they can still be overridden.
By splitting `dev_build=yes` engine developers could simply set it the `custom.py` and have those defaults set for all targets (maybe ignore it if target is "template-release"?).

akien-mga commented:
> - I reused the existing dev option to enable DEV_ENABLED and the default optimization levels and debug symbols, which means that it now sets all this (all can be overridden if specified on the command line): optimize=debug debug_symbols=yes verbose=yes warnings=extra werror=yes tests=yes + DEV_ENABLED
>   * "Forcing" all contributors to use verbose=yes and warnings=extra werror=yes may be considered too much. They're not fully forced as they can override it (from SCons, it's harder with the vsproj stuff, more on that below).
>     - I might change my mind there, I'm not fully happy with it. Especially since the warnings depend on the optimization level so we probably want to encourage contributors to occasionally build release templates with warnings=extra werror=yes and dev=yes was a convenient shortcut for it. It would just need a new name if we want to keep it separate.

Yeah I changed my mind, I think this was going a bit too far.

So it's now `dev_build=yes` to enable `DEV_ENABLED` and set default debug optimization level and debug symbols, and the old `dev=yes` convenient alias for `verbose=yes warnings=extra werror=yes tests=yes` is now `dev_mode=yes`.

So the recommend workflow for contributors is to build the editor with `scons dev_build=yes dev_mode=yes` but the last one can be left out if you don't like warnings or verbose mode.

akien-mga commented:
This should be more or less ready for review, removing the "WIP" flag.

It needs extensive testing on various OSes to make sure that it behaves like we want it.

Some remaining TODOs:
- TextServer gdextension code needs review, I'm not sure its code handling `env["target"]` will behave as expected.
- `vsproj=yes` should be tested and fixed to generate something useful for MSVC users. We probably don't need to generate "dev" variants for all targets, "editor" might be enough, maybe also "template-debug".
- See how to do similar changes in godot-cpp once we agree on the design.
- Update the documentation.

akien-mga commented:
This is now ready for merge and can get another pass of code/design review, and ideally a lot of user testing.

Notable change since the previous iteration, I made `dev_build=yes` default to `optimize=none` (`-O0`) instead of `optimize=debug` (`-Og`). This is more in line with the behavior before this PR as we didn't use `-Og` on any platform for `target=debug`. Early tests showed that `-Og` builds take significantly longer (to a pathological point where our CI build with GCC, editor, doubles, sanitizers, and debug symbols would take more than 3h and end up killed by GitHub Actions for filling up the disk).

I think it's best if we default to options which make the build fast, and let users who need more use specific flags. Nobody complained so far about the stacktrace quality of our previous `target=debug` setting (which used `-O0`) so we don't really need to move to `-Og`. It might optimize the editor speed somehow so benchmark build time + editor performance would be worth doing in a follow-up to possibly reconsider this.

I updated the OP and commit message with full details on the current options, so you can read that to know what this PR does in its current implementation.

aaronfranke commented:
By the way, is there a reason that this is using `-` instead of `_`? The old code had `release_debug`.

akien-mga commented:
> By the way, is there a reason that this is using `-` instead of `_`? The old code had `release_debug`.

Personal preference. This is the more commonly used format for command line arguments, e.g. `--some-option some-value`. I don't remember the exact circumstances but I've often faced situations where a terminal highlights spaces with an underscore, making them impossible to distinguish from actual underscore characters in the original string.

Since we use underscores for option names, I like that there is now a clear difference between option names and option values:
`scons target=template-debug debug_symbols=yes`. But I'm fine changing it back to all underscores if that's what most prefer.

Riteo commented:
@akien-mga IMO we should strive for consistency. I can see `dev_build`, `debug_symbols` and `release-debug` from just looking at the OP. Either we go all dashes or we go all underscores.

akien-mga commented:
Well I got it settled with a ~trial by combat~ quick 'n dirty poll among maintainers:
![image](https://user-images.githubusercontent.com/4701338/192150357-9b507dd2-9e48-4619-a219-b1ab887764c2.png)

aaronfranke commented:
> Do note that it implies binaries will be named `godot.linuxbsd.template_release.x86_64` which IMO looks bad, but it's not a big deal :P

That's actually the reason I suggested it, I thought `godot.linuxbsd.template-release.x86_64` looked bad :P

Bromeon commented:
I got an ICE (internal compiler error) on Windows in `editor/editor_feature_profile.cpp`, in two builds, but couldn't reproduce it in CI.

Compiled with `scons p=windows -j12 target=editor bits=64 debug_symbols=yes`. I guess this is not a valid config?
With `scons p=windows -j12 target=editor bits=64 dev_build=yes`, it works.

<details>

![grafik](https://user-images.githubusercontent.com/708488/192257160-0e9fe798-9d09-4ebd-a0b1-72a795b5f7b9.png)

</details>

akien-mga commented:
> Compiled with `scons p=windows -j12 target=editor bits=64 debug_symbols=yes`. I guess this is not a valid config?
> With `scons p=windows -j12 target=editor bits=64 dev_build=yes`, it works.

The `bits=64` part is obsolete, you can remove it (you'd use `arch` now to specify the architecture if cross-compiling).

Otherwise the difference between the two commands is that the first one would use `optimize=speed-trace` (`/O2` with MSVC) while the latter uses `optimize=none` (`/Od`).

Seems like MSVC is not happy generating debug symbols for an optimized build. If so you should have the same issue in the current `master` branch compiling with `scons p=windows -j12 tools=yes target=release_debug debug_symbols=yes` which would be equivalent.

Bromeon commented:
Thanks! I'll see if I can reproduce the same issue with master branch. 👍 

I generally find the new flags a bit confusing, because they are not really orthogonal. Many of them implicitly include other flags, or are mutually exclusive with others (but this is not validated by scons). This is at least the impression I have from your initial post in this PR. Do you think _only_ providing the low-level options would make scons too hard to use?

Is there something like a cheatsheet/doc which lists every options and groups them so it's easy to see what can go with what?

Bromeon commented:
Just to confirm: you were right that `scons p=windows -j12 tools=yes target=release_debug debug_symbols=yes` on `master` would also cause an ICE. 💥

neikeq commented:
> Many of them implicitly include other flags, or are mutually exclusive with others (but this is not validated by scons).

Maybe SCons should beade to print an error or a warning in this case.

akien-mga commented:
> > Many of them implicitly include other flags, or are mutually exclusive with others (but this is not validated by scons).
> 
> Maybe SCons should be made to print an error or a warning in this case.

There are no mutually exclusive flags currently (at least in what this PR implements). There are some combinations which are not super _useful_ (e.g. `target=release-template optimize=none debug_symbols=no`) but it's still a valid combination.

The MSVC ICE is a bug but this should be a valid combination too (and AFAIK it's the default if you want to build an optimized editor binary prior to this PR so I'm surprised by the ICE, you might just be running out of memory and it might pass with less CPU usage).

> Do you think only providing the low-level options would make scons too hard to use?

Yes, many users do custom builds and it's difficult enough for them to understand what the options are to build an optimized version of the editor, and release and debut templates. See the linked proposal for more details.

The new targets are helpers to set the right options out of the box. You don't have to specifying anything more than `scons target=...` to build one of the three targets in a way that's somewhat usable in production.

If we expose everything as low-level options then users would need to know:
- Production editor build: `editor=yes optimize=speed_trace debug_features=yes dev_features=no debug_symbols=no`
- Development editor build: `editor=yes optimize=none debug_features=yes dev_features=yes debug_symbols=yes`
- I leave the other combinations as an exercise to the reader ;)

I thought about exposing more of the low-level options (`editor`, `debug_features`) and high-level targets, but then you do get in a situation with a lot of invalid combinations that need to be handled. The current presets + `dev_build` expose all the combinations 99% of users would ever need.

akien-mga commented:
Here we go! As a reminder:

- Check the PR description to know how to change the way you compile Godot from now on (`master` only, if you also build `3.x` you should keep using the old `tools`/`target` there)
- Both the intermediate object files (`.o`, `.a`, `.obj`, `.lib`) and the Godot binary in bin are changing names, so you might want to clean your Git checkout to remove the old ones to save disk space, and adjust any scripts/config you might have that runs the Godot binary to use the new name.

Zireael07 commented:
How about putting that in documentation? IIRC there's a whole section on compiling with Scons...

akien-mga commented:
> How about putting that in documentation? IIRC there's a whole section on compiling with Scons...

Sure, but one thing at a time. See https://github.com/godotengine/godot/pull/66242#issuecomment-1256350127

akien-mga commented on methods.py:
```diff
@@ -779,16 +778,16 @@ def add_mode(
+                    f'bin\\godot.windows.{config}{dev}.{plat_id}{f".{name}" if name else ""}.exe'
+                    for config in ModuleConfigs.CONFIGURATIONS
                     for plat_id in ModuleConfigs.PLATFORM_IDS
+                    for dev in ModuleConfig.DEV_SUFFIX
```
Yes, feel free to make a PR if you can also validate that it actually works fine to build in VS.

akien-mga replied:
Fixed in 166df0896c.

EricEzaM commented:
Also cannot build vsproj:
```
[Initial build] scons: *** [godot.vcxproj] InternalError : Sizes of 'buildtarget' and 'variant' lists must be the same.
Traceback (most recent call last):
  File "C:\Users\Eric\AppData\Local\Programs\Python\Python38\lib\site-packages\SCons\Action.py", line 1318, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "C:\Users\Eric\AppData\Local\Programs\Python\Python38\lib\site-packages\SCons\Tool\msvs.py", line 1845, in GenerateProject
    GenerateDSP(dspfile, source, env)
  File "C:\Users\Eric\AppData\Local\Programs\Python\Python38\lib\site-packages\SCons\Tool\msvs.py", line 1795, in GenerateDSP
    g = _GenerateV10DSP(dspfile, source, env)
  File "C:\Users\Eric\AppData\Local\Programs\Python\Python38\lib\site-packages\SCons\Tool\msvs.py", line 1196, in __init__
    _DSPGenerator.__init__(self, dspfile, source, env)
  File "C:\Users\Eric\AppData\Local\Programs\Python\Python38\lib\site-packages\SCons\Tool\msvs.py", line 448, in __init__
    raise SCons.Errors.InternalError("Sizes of 'buildtarget' and 'variant' lists must be the same.")
SCons.Errors.InternalError: Sizes of 'buildtarget' and 'variant' lists must be the same.
scons: building terminated because of errors.
```

akien-mga commented:
@EricEzaM Opened an issue to keep track of it: #66664.
